rewriter/src/main/java/com/facebook/coresql/rewriter/ApproxPercentileRewriter.java

```diff
@@ -0,0 +1,182 @@
+    }
+
+    @Override
+    public Optional<RewriteResult> rewrite()
+    {
+        if (!approxPercentileRewritePatternIsPresent()) {
+            return Optional.empty();
+        }
+        String rewrittenSql = unparse(root, this);
+        return Optional.of(new RewriteResult(REWRITE_NAME, rewrittenSql));
+    }
+
+    @Override
+    public void visit(FunctionCall node, Void data)
+    {
+        if (canApplyApproxPercentileRewrite(node)) {
+            applyApproxPercentileRewrite(node);
+        }
+        else {
+            defaultVisit(node, data);
+        }
+    }
+
+    private boolean approxPercentileRewritePatternIsPresent()
+    {
+        return firstArgToApproxPercentileNode.keySet().stream()
+                .anyMatch(key -> firstArgToApproxPercentileNode.get(key).size() >= 2);
+    }
+
+    private boolean canApplyApproxPercentileRewrite(AstNode node)
+    {
+        return firstArgToApproxPercentileNode.containsValue(node) && firstArgToApproxPercentileNode.get(unparse(getNthArgument(node, 0).get()).trim()).size() >= 2;
+    }
+
+    /**
+     * Generates a rewritten version of the current subtree.
+     *
+     * @param node The function call node we're rewriting
+     */
+    private void applyApproxPercentileRewrite(AstNode node)
+    {
+        // First, unparse up to the node's first child. This ensures we don't miss any special tokens
+        unparseUpto((AstNode) node.jjtGetChild(0));
+        // Then, add the rewritten version to the rewriter's result object (i.e. stringBuilder)
+        String firstArg = unparse(getNthArgument(node, 0).get()).trim();
+        String secondArg = unparse(getNthArgument(node, 1).get()).trim();
+        Formatter formatter = new Formatter(stringBuilder);
+        formatter.format(REPLACEMENT_FORMAT, firstArg, firstArgToPercentiles.get(firstArg), getIndexOfPercentile(firstArg, secondArg));
+        // Lastly, move to end of this node -- we've already added a rewritten version of it to the result, so we don't need to process it further
+        moveToEndOfNode(node);
+    }
+
+    private int getIndexOfPercentile(String firstArg, String secondArg)
+    {
+        return firstArgToPercentiles.get(firstArg).indexOf(secondArg) + ZERO_INDEXING_OFFSET;
+    }
+
+    private static Optional<AstNode> getNthArgument(AstNode node, int n)
+    {
+        Optional<AstNode> argList = Optional.ofNullable(node.GetFirstChildOfKind(JJTARGUMENTLIST));
+        if (!argList.isPresent() || argList.get().jjtGetNumChildren() < n) {
+            return Optional.empty();
+        }
+        AstNode nthArg = (AstNode) argList.get().jjtGetChild(n);
+        return Optional.of(nthArg);
+    }
+
+    private static boolean hasUnsignedLiteralSecondArg(AstNode node)
+    {
+        return getNthArgument(node, 1).filter(astNode -> astNode.getId() == JJTUNSIGNEDNUMERICLITERAL).isPresent();
+    }
+
+    private static boolean isApproxPercentileNode(AstNode node)
```

Comment:
Is it needed here? You should move it to the pattern matcher class.

Reply:
Will be moved to NodeUtils in a future PR.